prow/spyglass/gcsartifact_fetcher.go

```diff
@@ -0,0 +1,256 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spyglass
+
+import (
+	"crypto/tls"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/spyglass/viewers"
+)
+
+// A fetcher for a GCS client
+type GCSArtifactFetcher struct {
+	client      *storage.Client
+	xmlEndpoint string
+	withTLS     bool
+}
+
+// A location in GCS where Prow job-specific artifacts are stored. This implementation assumes
+// Prow's native GCS upload format (treating GCS keys as a directory structure), and is not
+// intended to support arbitrary GCS bucket upload formats.
+type GCSJobSource struct {
+	source     string
+	linkPrefix string
+	bucket     string
+	jobPath    string
+	jobName    string
+	buildID    string
+}
+
+// NewGCSArtifactFetcher creates a new ArtifactFetcher with a real GCS Client
+func NewGCSArtifactFetcher(c *storage.Client, xmlEndpoint string, tls bool) *GCSArtifactFetcher {
+	return &GCSArtifactFetcher{
+		client:      c,
+		xmlEndpoint: xmlEndpoint,
+		withTLS:     tls,
+	}
+}
+
+// NewGCSJobSource creates a new GCSJobSource from a given bucket and jobPath
+func NewGCSJobSource(src string) *GCSJobSource {
+	linkPrefix := "gs://"
+	noPrefixSrc := strings.TrimPrefix(src, linkPrefix)
+	if !strings.HasSuffix(noPrefixSrc, "/") { // Cleaning up path
+		noPrefixSrc += "/"
+	}
+	tokens := strings.FieldsFunc(noPrefixSrc, func(c rune) bool { return c == '/' })
+	bucket := tokens[0]
+	buildID := tokens[len(tokens)-1]
+	name := tokens[len(tokens)-2]
+	jobPath := strings.TrimPrefix(noPrefixSrc, bucket+"/") // Extra / is not part of prefix, only necessary for URI
+	return &GCSJobSource{
+		source:     src,
+		linkPrefix: linkPrefix,
+		bucket:     bucket,
+		jobPath:    jobPath,
+		jobName:    name,
+		buildID:    buildID,
+	}
+}
+
+// isGCSSource recognizes whether a source string references a GCS bucket
+func isGCSSource(src string) bool {
```

Comment:
I still feel like this is a bad design and should be reconsidered right now before any more work is done ...

Reply:
What is bad about this design?

Reply:
You take structured data -- artifact type, names, fields, stringly type it, then parse everything back out later. The impl invents a new URL format with a leading `prowjob` to identify pod logs... we should just use types.

Reply:
(Also I've written this review like ten times now and I'm a little tired of rewriting it on every PR)

Reply:
We're using string keys, because the keys are strings... they come from a URL segment. Go does not have string enums. We're not trying to shove some format like JSON into the URL.